### DISCUSSION TITLE
Failed to build NXP mimxrt1060_evk board

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#75037] [Q&A] [answered]
Hey, 

I recently synced my Zephyr project with the latest main branch on ee451e5cf050285b850fdef8e6afaec666994928 commit and ran `west update`, and now I'm encountering build failures. 

Compilation Output:
```
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: C:/Users/ofirs/bsp
-- CMake version: 3.29.1
-- Cache files will be written to: C:/Users/ofirs/bsp/zephyr/.cache
-- Zephyr version: 3.7.0-rc1 (C:/Users/ofirs/bsp/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: mimxrt1060_evk, Revision: qspi, qualifiers: mimxrt1062
-- Found toolchain: gnuarmemb (C:\Users\ofirs\bsp\gcc-arm-none-eabi-10.3-2021.10)
-- Found BOARD.dts: C:/Users/ofirs/bsp/zephyr/boards/nxp/mimxrt1060_evk/mimxrt1060_evk.dts
-- Found devicetree overlay: C:/Users/ofirs/bsp/zephyr/boards/nxp/mimxrt1060_evk/mimxrt1060_evk_mimxrt1062_qspi.overlay
-- Generated zephyr.dts: C:/Users/ofirs/bsp/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: C:/Users/ofirs/bsp/build/zephyr/include/generated/zephyr/devicetree_generated.h

-- Including generated dts.cmake file: C:/Users/ofirs/bsp/build/zephyr/dts.cmake
Parsing C:/Users/ofirs/bsp/zephyr/Kconfig
Loaded configuration 'C:/Users/ofirs/bsp/zephyr/boards/nxp/mimxrt1060_evk/mimxrt1060_evk_defconfig'
Merged configuration 'C:/Users/ofirs/bsp/project_configuration/prj.conf'
Configuration saved to 'C:/Users/ofirs/bsp/build/zephyr/.config'
Kconfig header saved to 'C:/Users/ofirs/bsp/build/zephyr/include/generated/zephyr/autoconf.h'
Load components for MIMXRT1062:
driver_common component is included.
driver_reset component is included.
device_CMSIS component is included.
CMSIS_Include_core_cm component is included.
device_system component is included.
driver_trng component is included.
driver_enet component is included.
driver_lpi2c component is included.
driver_lpspi component is included.
driver_lpuart component is included.
driver_igpio component is included.
driver_flexspi component is included.
driver_cache_armv7_m7 component is included.
driver_ocotp component is included.
-- Configuring done (8.6s)
-- Generating done (1.2s)
-- Build files have been written to: C:/Users/ofirs/bsp/build
←[92m-- west build: building application
[355/360] Linking C executable zephyr\zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map C:/Users/ofirs/bsp/build/zephyr/zephyr_pre0.map
C:\Windows\system32\cmd.exe /C "cd . && C:\Users\ofirs\bsp\gcc-arm-none-eabi-10.3-2021.10\bin\arm-none-eabi-gcc.exe  --specs=nosys.specs   -gdwarf-4 @CMakeFiles\zephyr_pre0.rsp -o zephyr\zephyr_pre0.elf && C:\Windows\system32\cmd.exe /C "cd /D C:\Users\ofirs\bsp\build\zephyr && "C:\Program Files\CMake\bin\cmake.exe" -E true""
c:/users/ofirs/bsp/gcc-arm-none-eabi-10.3-2021.10/bin/../lib/gcc/arm-none-eabi/10.3.1/../../../../arm-none-eabi/bin/ld.bfd.exe: C:/Users/ofirs/bsp/gcc-arm-none-eabi-10.3-2021.10/arm-none-eabi/lib/thumb/v7e-m/nofp\libc.a(lib_a-fini.o): in function `__libc_fini_array':
fini.c:(.text.__libc_fini_array+0x20): undefined reference to `_fini'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\ofirs\bsp\build'
```

It seems like a linker error. 

Your help would be much appreciated.

Best regards,
Ofir

**Comment by DerekSnell** [comment]
Hi @ofirshe ,
Searching the web, I see [references](https://community.platformio.org/t/linker-error-with-config-net-tcp-y/24143/9) that this linker error is related to the C standard library.  Are you using C++ in your app?

One change that comes to mind last year is Zephyr changed the default library from Newlib to Picolib, see https://github.com/zephyrproject-rtos/zephyr/pull/57340.  Perhaps the library building with your app now changed, or the settings for that library changed.  I would compare the Kconfig generated file build/zephyr/.config from before and after, search for the LIBC Kconfigs, and see if those have changed.

Let us know what you find

**Reply by ofirshe** [reply, marked as answer]
Hi @DerekSnell,

I managed to compile successfully after removing these three configs:
```
CONFIG_NEWLIB_LIBC=y
CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
CONFIG_NEWLIB_LIBC_NANO=n
```
These were previously used for floating point support but are no longer necessary.

I see that Picolib is enabled by default, as expected. Would you recommend using this library for the RT1060 controller, or do you have another suggestion?

Additionally, I've observed that the compilation now takes significantly longer [1303/1303] compared to before [355/360].

Thanks.

**Comment by DerekSnell** [comment]
Hi @ofirshe ,
I am glad you resolved the build issue.

> I see that Picolib is enabled by default, as expected. Would you recommend using this library for the RT1060 controller, or do you have another suggestion?

Zephyr integrates multiple standard libraries to be flexible for the application.  The best one to use will depend on the application.  I am not an expert on the libraries.  My understanding is that Picolibc is basically a subset of Newlib, this is a [summary](https://chromium.googlesource.com/chromiumos/third_party/zephyr/picolibc/+/b92edfda8ac6853772d87cadaeeeaa21b78609b6/README.md) I found.  Picolibc targets smaller resource-constrained devices, and should consume less memory.  But it must not include all the features of Newlib.

Best regards

**Reply by ofirshe** [reply]
Thanks for the clarification !
